internal/testrunner/runners/system/servicedeployer/terraform_test.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package servicedeployer
+
+import (
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFindPolicyTemplateForInput(t *testing.T) {
+
+	var testCases = []struct {
+		testName      string
+		err           string
+		ctxt          ServiceContext
+		runId         string
+		content       []byte
+		expectedProps map[string]interface{}
+	}{
+		{
+			testName: "single_value_output",
+			runId:    "99999",
+			ctxt: ServiceContext{
+				Test: struct{ RunID string }{"99999"},
+			},
+			content: []byte(
+				`{
+				"queue_url": {
+				  "sensitive": false,
+				  "type": "string",
+				  "value": "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId"
+				}
+			   }`,
+			),
+			expectedProps: map[string]interface{}{
+				"TF_OUTPUT_queue_url": "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId",
+			},
+		},
+		{
+			testName: "multiple_value_output",
+			runId:    "23465",
+			ctxt: ServiceContext{
+				Test: struct{ RunID string }{"23465"},
+			},
+			content: []byte(
+				`{
+				"queue_url": {
+				  "sensitive": false,
+				  "type": "string",
+				  "value": "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId"
+				},
+				"instance_id": {
+					"sensitive": false,
+					"type": "string",
+					"value": "some-random-id"
+				  }
+			   }`,
+			),
+			expectedProps: map[string]interface{}{
+				"TF_OUTPUT_queue_url":   "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId",
+				"TF_OUTPUT_instance_id": "some-random-id",
+			},
+		},
+		{
+			testName: "complex_value_output",
+			runId:    "078907890",
+			ctxt: ServiceContext{
+				Test: struct{ RunID string }{"078907890"},
+			},
+			content: []byte(
+				`{
+					"queue_url": {
+					  "sensitive": false,
+					  "type": "string",
+					  "value": "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId"
+					},
+					"triangle_output": {
+					  "sensitive": false,
+					  "type": [
+						"object",
+						{
+						  "description": "string",
+						  "s_one": "number",
+						  "s_three": "number",
+						  "s_two": "number"
+						}
+					  ],
+					  "value": {
+						"description": "this is a triangle",
+						"s_one": 1,
+						"s_three": 2.5,
+						"s_two": 2.5
+					  }
+					}
+				}`,
+			),
+			expectedProps: map[string]interface{}{
+				"TF_OUTPUT_queue_url":                   "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId",
+				"TF_OUTPUT_triangle_output.s_one":       1.0,
```

Comment:
Why are these expected to be flattened? Handlebars supports[^1][^2] addressing nested values through its path expressions. So in theory if you write `{{TF_OUTPUT_triangle_output.s_one}}` then it should return the nested value without needed to store it as flattened. If Handlebars is not working as expected then I would like to try to solve that issue.

[^1]: https://handlebarsjs.com/guide/expressions.html#path-expressions
[^2]: https://github.com/aymerick/raymond/blob/b565731e1464263de0bda75f2e45d97b54b60110/handlebars/basic_test.go#L385-L386

This is what I was thinking...

```diff
diff --git a/internal/testrunner/runners/system/servicedeployer/terraform.go b/internal/testrunner/runners/system/servicedeployer/terraform.go
index 07322e5..5dd2911 100644
--- a/internal/testrunner/runners/system/servicedeployer/terraform.go
+++ b/internal/testrunner/runners/system/servicedeployer/terraform.go
@@ -14,7 +14,6 @@ import (
 
 	"github.com/pkg/errors"
 
-	"github.com/elastic/elastic-agent-libs/mapstr"
 	"github.com/elastic/go-resource"
 
 	"github.com/elastic/elastic-package/internal/compose"
@@ -59,25 +58,24 @@ func addTerraformOutputs(outCtxt ServiceContext) error {
 	// Unmarshall the data into `terraformOutputs`
 	logger.Debug("Unmarshalling terraform output json")
 
-	var terraformOutputs map[string]interface{}
-	err = json.Unmarshal(content, &terraformOutputs)
-	if err != nil {
-		return fmt.Errorf("error during json Unmarshal %w", err)
+	// https://github.com/hashicorp/terraform/blob/v1.4.6/internal/command/views/output.go#L217-L222
+	type OutputMeta struct {
+		Value interface{} `json:"value"`
 	}
+	var outputMetas map[string]OutputMeta
 
-	// Flatten if there are any objects in the output json
-	outputs := (mapstr.M).Flatten(terraformOutputs)
-	customProps := outCtxt.CustomProperties
+	if err = json.Unmarshal(content, &outputMetas); err != nil {
+		return fmt.Errorf("error during json Unmarshal %w", err)
+	}
 
-	// Considering only values from the output json
-	// https://github.com/hashicorp/terraform/blob/v1.4.6/internal/command/views/output.go#L217-L222
-	// Add the terraform outputs to custom properties with the key appeneded by "TF_OUTPUT_"
-	for k, v := range outputs {
-		if strings.Contains(k, ".value") {
-			newKey := strings.Replace(k, ".value", "", 1)
-			customProps[terraformOutputPrefix+newKey] = v
-		}
+	if len(outputMetas) > 0 && outCtxt.CustomProperties == nil {
+		outCtxt.CustomProperties = make(map[string]any, len(outputMetas))
 	}
+	// Prefix variables names with TF_OUTPUT_
+	for k, om := range outputMetas {
+		outCtxt.CustomProperties[terraformOutputPrefix+k] = om.Value
+	}
+
 	return nil
 }
 
diff --git a/internal/testrunner/runners/system/servicedeployer/terraform_test.go b/internal/testrunner/runners/system/servicedeployer/terraform_test.go
index 435038d..117c162 100644
--- a/internal/testrunner/runners/system/servicedeployer/terraform_test.go
+++ b/internal/testrunner/runners/system/servicedeployer/terraform_test.go
@@ -99,11 +99,13 @@ func TestFindPolicyTemplateForInput(t *testing.T) {
 				}`,
 			),
 			expectedProps: map[string]interface{}{
-				"TF_OUTPUT_queue_url":                   "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId",
-				"TF_OUTPUT_triangle_output.s_one":       1.0,
-				"TF_OUTPUT_triangle_output.s_two":       2.5,
-				"TF_OUTPUT_triangle_output.s_three":     2.5,
-				"TF_OUTPUT_triangle_output.description": "this is a triangle",
+				"TF_OUTPUT_queue_url": "https://sqs.us-east-1.amazonaws.com/1234654/elastic-package-aws-logs-queue-someId",
+				"TF_OUTPUT_triangle_output": map[string]any{
+					"description": "this is a triangle",
+					"s_one":       1.0,
+					"s_two":       2.5,
+					"s_three":     2.5,
+				},
 			},
 		},
 	}
```

Reply:
Looks like Handlebars is handling this already.. Do not have to Flatten stuff